src/fv/share/tdyugrid.c

```diff
@@ -0,0 +1,1209 @@
+#include <petscdm.h>
+#include <petscdmda.h>
+#include <petscviewerhdf5.h>
+#include <private/tdyugridimpl.h>
+#include <private/tdymemoryimpl.h>
+#include <private/tdyutils.h>
+
+static PetscInt vertex_separator = -777;
+static PetscInt dual_separator = -888;
+static PetscInt cell_separator = -999999;
+
+/// Allocates memory and initializes a TDyUGrid struct
+///
+/// @param [out] ugrid A TDyUGrid struct
+///
+/// @returns 0 on success, or a non-zero error code on failure
+PetscErrorCode TDyUGridCreate(TDyUGrid **ugrid) {
+
+  PetscErrorCode ierr;
+
+  ierr = PetscCalloc(sizeof(TDyUGrid), ugrid); CHKERRQ(ierr);
```

Comment:
The difference between `malloc` and `calloc` (and `PetscMalloc` and `PetscCalloc`) is that `calloc` zeroes out all memory that is allocated, which makes lines 23-56 unnecessary. (`malloc`, by contrast, just hands you back a chunk of uninitialized memory.)